### PR TITLE
Authentication route using jwt tokens and password salt

### LINKED ISSUE #9: Server Authentication Middleware + Local strategy
1. Choose an identity provider (AD, Auth0)
2. Implement the authentication and authorization flow machine-to-machine for the server

### PR DESCRIPTION
**Pull request by YichiZ**
- Create auth route for authentication (login / register / me) using jwt tokens
- Added cors
- Made Email unique
- Salt password before saving to db
- Authorization middleware

Closes: YichiZ/coachbox#9

### CODE REVIEW

**Comment by YichiZ** [comment]
This PR contain authentication and authorization

File: middleware/Authorize.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const jwt = require('jsonwebtoken');
+
+module.exports = (req, res, next) => {
+    var token = req.headers['authorization'].split(' ')[1];
```
Comment (TheAlexRobert) [inline middleware/Authorize.js]: Can you explain this line "req.headers['authorization'].split(' ')[1]"

File: routes/auth.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+const router = require('express').Router();
+const User = require('../models/User');
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+
+router.post('/register', async (req, res) => {
+    const saltRounds = 10;
```
Comment (TheAlexRobert) [inline routes/auth.js:8]: why saltRounds=10?

File: routes/auth.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        res.status(200).send(decoded);
+    });
+});
+
+router.post('/login', async (req, res) => {
+    const user = await User.findOne({ email: req.body.email})
+        .exec()
+        .catch(err => res.send(500).send(err));
+
+    const isPasswordValid = await bcrypt.compare(req.body.password, user.password);
```
Comment (TheAlexRobert) [inline routes/auth.js:34]: good, so we never actually store passwords right? It directly compares hash if I understand correctly

**Review by TheAlexRobert — CHANGES_REQUESTED** [review]
Looks good, can you add screenshots whenever you do a PR (Pull Request). Just to show that it works.  And could you have a look at the comments I left.

File: routes/auth.js (hunk 2 above)
Comment (YichiZ) [reply routes/auth.js:8]: It was the value recommended online. From what I understand: 
the lower it is, the more vulnerable the password becomes
The higher it is, the more time it takes to generate the hashed password

File: routes/auth.js (hunk 3 above)
Comment (YichiZ) [reply routes/auth.js:34]: Yes exactly. This adds a layer of security

File: middleware/Authorize.js (hunk 1 above)
Comment (YichiZ) [reply middleware/Authorize.js]: The header is an object that look like this:
{
    "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
}

Comment (YichiZ) [reply middleware/Authorize.js:4]: The second part is the jwt token needed to authorize the user

Comment (YichiZ) [reply middleware/Authorize.js]: We need the second part (jwt token) to verify the user

**Comment by YichiZ** [comment]
Yes of course, I'll send you some screenshot tonight

**Comment by YichiZ** [comment]
Send a request with token: 
image

How the request look like with a debugger
image

File: routes/auth.js (hunk 2 above)
Comment (TheAlexRobert) [reply routes/auth.js:8]: Perfect, I trust your judgment, we can always change it au pire

File: middleware/Authorize.js (hunk 1 above)
Comment (TheAlexRobert) [reply middleware/Authorize.js]: Excellent, I understand now

**Review by TheAlexRobert — APPROVED** [review]
Great work, thank you for the screenshots

**Comment by YichiZ** [comment]
Excellent, yes we can definitely change in the future. There are multiple strategies for authentication. Passport.js is a big one but I wasn't sure how to use it properly

**Pull request merged**